### PR TITLE
Add a Physics class for linearFV turbulence modelling

### PR DESCRIPTION
**Pull request by GiudGiud**
refs https://github.com/idaholab/moose/issues/29175

Testing includes the traditional other Physics we like to use with turbulence

### CODE REVIEW

File: framework/src/physics/PhysicsBase.C [hunk 1]
```diff
@@ -540,6 +540,16 @@ PhysicsBase::checkBlockRestrictionIdentical(const std::string & object_name,
...
 }
 
+bool
+PhysicsBase::hasBlocks(const std::vector<SubdomainName> & blocks) const
+{
+  mooseAssert(_blocks.size(), "hasBlocks called before blocks were initialized");
+  return std::all_of(blocks.begin(),
+                     blocks.end(),
+                     [this](const SubdomainName & block)
+                     { return std::find(_blocks.begin(), _blocks.end(), block) != _blocks.end(); });
```
Comment (lindsayad) [inline framework/src/physics/PhysicsBase.C:557]: If you have `_blocks` sorted and unique, then you could use `std::binary_search`

Comment (GiudGiud) [reply framework/src/physics/PhysicsBase.C:557]: unique yes.
sorted is user-sorting (unfortunately maybe)

File: modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+!syntax description /Physics/NavierStokes/TurbulenceSegregated/WCNSLinearFVTurbulencePhysics
+
+## K-Epsilon turbulence model
+
+The `WCNSLinearFVTurbulencePhysics` can be set to create the standard k-epsilon two-equation model.
+
+The turbulent viscosity is then computed with a [kEpsilonViscosityAux.md].
+
+
+The k equation is created with:
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md]:
```suggestion
The k (turbulent kinetic energy, TKE) equation is created with:
```

File: modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+- a [LinearFVTimeDerivative.md] for the time derivative if simulating a transient
+- a [LinearFVTurbulentAdvection.md] for the turbulent kinetic energy advection term
+- a [LinearFVTurbulentDiffusion.md] for the turbulent kinetic energy diffusion term
+- a [LinearFVTKESourceSink.md] for the turbulent kinetic energy source and dissipation (sink) terms
+
+
+The epsilon equation is created with:
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md]:
```suggestion
The epsilon (turbulent kinetic energy dissipation, TKED) equation is created with:
```

File: modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+The boundary conditions are not set in this object for the `TKE` and `TKED` variables, as they
+are computed by the wall-functions in the relevant kernels. A boundary condition is set for the turbulent
+viscosity when using an auxiliary variable, with a [LinearFVTurbulentViscosityWallFunctionBC.md].
+
+
+## Coupling with other Physics
+
+A turbulence model can be added to a heat advection solve by using both a `WCNSLinearFVTurbulencePhysics` and a [WCNSLinearFVFluidHeatTransferPhysics.md].
+The following input performs this coupling for incompressible flow for the standard k-epsilon turbulence model in a 2D channel.
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md:36]: Maybe a warning that the wall functions for temperature haven't been implemented yet?

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+InputParameters
+WCNSFVTurbulencePhysicsBase::validParams()
+{
+  InputParameters params = NavierStokesPhysicsBase::validParams();
+  params += WCNSFVCoupledAdvectionPhysicsHelper::validParams();
+  params.addClassDescription(
+      "Define a turbulence model for a incompressible or weakly-compressible Navier Stokes "
+      "flow with a finite volume discretization");
+
+  MooseEnum turbulence_type("mixing-length k-epsilon none", "none");
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:30]: We don't have mixing length for linear FV.

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+  // Boundary parameters
+  params.addParam<bool>("bulk_wall_treatment", true, "Whether to treat the wall cell as bulk");
+  MooseEnum wall_treatment("eq_newton eq_incremental eq_linearized neq", "neq");
+  params.addParam<MooseEnum>("wall_treatment_eps",
+                             wall_treatment,
+                             "The method used for computing the epsilon wall functions and the "
+                             "turbulence viscosity wall functions");
+  params.addParam<MooseEnum>("wall_treatment_T",
+                             wall_treatment,
+                             "The method used for computing the temperature wall functions");
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:76]: We don't have these for linear FV.

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+  params.transferParam<Real>(INSFVTurbulentViscosityWallFunction::validParams(), "C_mu");
+
+  // K-Epsilon numerical scheme parameters
+  MooseEnum face_interpol_types("average skewness-corrected", "average");
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:80]: We dont have skewness corrected for linear FV

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+                             "face (separate from the advected quantity interpolation).");
+  params.addParam<MooseEnum>("tke_advection_interpolation",
+                             adv_interpol_types,
+                             "The numerical scheme to interpolate the TKE to the "
+                             "face when in the advection kernel.");
+  params.addParam<bool>(
+      "tke_two_term_bc_expansion",
+      false,
+      "If a two-term Taylor expansion is needed for the determination of the boundary values"
+      "of the turbulent kinetic energy.");
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:94]: We don't support this for linear FV, same for TKED

Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:94]: Well, technically people could do it through some of the boundary conditions.

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+        params.set<std::vector<std::string>>("functor_names") = {_turbulent_viscosity_name};
+
+      params.set<std::string>("expression") =
+          _flow_equations_physics->dynamicViscosityName() + " + " + _turbulent_viscosity_name;
+      params.set<std::string>("property_name") = NS::mu_eff;
+      getProblem().addMaterial(object_type, prefix() + "effective_viscosity", params);
+    }
+    if (!getParam<bool>("mu_t_as_aux_variable"))
+    {
+      // TODO: suppress this for linearFV
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C]: Yes pls.

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+#include "NSFVBase.h"
+
+registerWCNSFVTurbulenceBaseTasks("NavierStokesApp", WCNSLinearFVTurbulencePhysics);
+
+InputParameters
+WCNSLinearFVTurbulencePhysics::validParams()
+{
+  InputParameters params = WCNSFVTurbulencePhysicsBase::validParams();
+  params.addClassDescription(
+      "Define a turbulence model for a incompressible or weakly-compressible Navier Stokes "
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]:
```suggestion
      "Define a turbulence model for an incompressible or weakly-compressible Navier Stokes "
```

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+{
+  if (_turbulence_model == "k-epsilon")
+    getProblem().needSolutionState(1, Moose::SolutionIterationType::Nonlinear);
+}
+
+void
+WCNSLinearFVTurbulencePhysics::addSolverVariables()
+{
+  if (_turbulence_model == "mixing-length" || _turbulence_model == "none")
+    return;
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:96]: Is there an error somewhere if we are doing mixing length with linear?

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+
+void
+WCNSLinearFVTurbulencePhysics::addFluidEnergyTurbulenceKernels()
+{
+  if (_turbulence_model == "k-epsilon")
+  {
+    // Simpler to add the kernel in the fluid heat transfer physics with k + k_t instead of two
+    // kernels one with k, one with k_t
+  }
+}
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]:
```suggestion
```

Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: You can add the comment in the header.

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+  const std::string kernel_type = "LinearFVTimeDerivative";
+  InputParameters params = getFactory().getValidParams(kernel_type);
+  assignBlocks(params, _blocks);
+
+  params.set<LinearVariableName>("variable") = _tke_name;
+  if (shouldCreateTimeDerivative(_tke_name, _blocks, /*error if already defined*/ false))
+    getProblem().addLinearFVKernel(kernel_type, prefix() + "tke_time", params);
+  params.set<LinearVariableName>("variable") = _tked_name;
+  if (shouldCreateTimeDerivative(_tked_name, _blocks, /*error if already defined*/ false))
+    getProblem().addLinearFVKernel(kernel_type, prefix() + "tked_time", params);
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:173]: We need a rho multiplier on k and epsilon in the time derivative.

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+    params.set<LinearVariableName>("variable") = _tke_name;
+    params.set<MooseFunctorName>("diffusion_coeff") = _turbulent_viscosity_name;
+    params.set<MooseFunctorName>("scaling_coeff") = getParam<MooseFunctorName>("sigma_k");
+    getProblem().addLinearFVKernel(kernel_type, prefix() + "tke_diffusion_mu_turb", params);
+
+    params.set<std::vector<BoundaryName>>("walls") = _turbulence_walls;
+    params.set<LinearVariableName>("variable") = _tked_name;
+    params.set<MooseFunctorName>("scaling_coeff") = getParam<MooseFunctorName>("sigma_eps");
+    getProblem().addLinearFVKernel(kernel_type, prefix() + "tked_diffusion_mu_turb", params);
+  }
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: We should have only one block `mu + mu_t/sigma`. Linear FV doesn't work very well with the definition of multiple kernels for the same physics. If you define a flux BC it will be added twice because of this.

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv [hunk 15]
```diff
@@ -0,0 +1,5 @@
+TKE,TKED,id,pressure,vel_x,vel_y,x,y,z
+0.0025877697461348,7.3889023714783e-05,0,0.34151550458492,1.040915708239,-0.00060794749797394,12.5,0.0001,0
+0.0022716271031286,0.00010389410436847,25,0.19805242257515,1.1031328293788,-0.00018981906625752,37.5,0.0001,0
+0.0024687725070957,0.00013918985689074,50,0.11414494135339,1.1159091029636,-9.0708639485251e-05,62.5,0.0001,0
+0.0027809910727729,0.00016717995137461,75,0.03957707667781,1.1166179183553,-2.7900067360663e-05,87.5,0.0001,0
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv]: Would be nice if we could just get the same results as the kernel-based model.

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests [hunk 16]
```diff
@@ -11,13 +11,67 @@
...
     requirement = 'For the same Reynolds number but scaled density over viscosity ratio, the system shall be able to yield the same results than a benchmark case for a channel flow.'
   []
+  [channel_ERCOFTAC_linear_physics]
+    design = 'WCNSLinearFVFlowPhysics.md WCNSLinearFVTurbulencePhysics.md'
+    issues = '#29846 #29175'
+    type = 'CSVDiff'
+    input = 'channel_ERCOFTAC-physics.i'
+    # use same gold files
+    prereq = 'channel_ERCOFTAC_linear'
+    # Physics includes deviatoric term with dynamic viscosity, regular input does not
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests:47]: Wow, how big is the difference because of this?

**Review by grmnptr — CHANGES_REQUESTED** [review]
Looks nice, just a few questions!

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C (hunk 5 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:30]: I ll add an error

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C (hunk 6 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:76]: suppressed

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C (hunk 7 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:80]: suppressed

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C (hunk 8 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:94]: already suppressed
  params.suppressParameter<bool>("tke_two_term_bc_expansion");
  params.suppressParameter<bool>("tked_two_term_bc_expansion");

File: modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C (hunk 9 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C]: already done

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 11 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:96]: added one

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: ugh this is what we do left and right though

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 13 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:173]: good catch

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: like both cases in the flow test bed I think

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv (hunk 15 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv]: we would need to add a knob to turn off the deviatoric term

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: see https://github.inl.gov/idaholab/open_pronghorn/pull/21/files#diff-c40e8affb07ecc7318dd6d00ff1afd13fa63e298767cbe1a38652cf3a7b38b76

Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: so if we use an auxkernel for mu_eff = mu_t + mu I think we would not use the boundary condition

Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: so we would need to write a new BC for this ?

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests (hunk 16 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests:47]: not huge
```
test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics: ERROR: In file channel_ERCOFTAC-physics_out_line_center_channel_0001.csv: The values in column "vel_y" don't match. 
test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics: 	relative diff:   -6.077e-04 ~ -6.079e-04 = 4.390e-04 (4.390e-04)
then upping the tol
test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics: ERROR: In file channel_ERCOFTAC-physics_out_line_center_channel_0001.csv: The values in column "vel_y" don't match. 
test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics: 	relative diff:   -9.053e-05 ~ -9.071e-05 = 1.993e-03 (1.993e-03)
```
thn it s ok with 3.5e-3 rel_err

do you prefer a tolerance?

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv (hunk 15 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv]: actually we already have one
```
  params.addParam<bool>(
      "include_deviatoric_stress",
      false,
      "Whether to include the full expansion (the transposed term as well) of the stress tensor");
```

so prefer that? a test with the wrong setup but no extra gold files?

Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv]: the option could be added from the tests along with a note saying "dont do it"

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: ok so we go the functor material route

Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: ugh it diffs

any ideas?

Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: Interesting. Which one diffs? And how much?

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv (hunk 15 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv]: Okay, I'd prefer having the deviatoric term.

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests (hunk 16 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests:47]: No, we are good on this!

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: all 3. a little bit

Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: I suspect once all the patches in the other PRs are merged it will be fine

File: modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md (hunk 4 above)
Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTurbulencePhysics.md:36]: done

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: nope. It still diffs. 
I found one culprit, we have a mistake in the turbulent diffusion kernel

File: modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv (hunk 15 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv]: I ll go the consistent route actually, because the deviatoric stress term has been added in that test now

File: modules/navier_stokes/src/functormaterials/FunctorEffectiveDynamicViscosity.C [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+      "property_name", "Name of the functor for the effective dynamic viscosity");
+  params.addRequiredParam<MooseFunctorName>(NS::mu, "Name of the dynamic viscosity functor");
+  params.addRequiredParam<MooseFunctorName>(NS::mu_t, "Name of the turbulent viscosity functor");
+  params.addRequiredParam<MooseFunctorName>(NS::mu_t + "_inverse_factor",
+                                            "Factor dividing the turbulent viscosity functor");
+  params.addParam<Real>(NS::mu_t + "_extra_inverse_factor",
+                        1.,
+                        "Additional factor dividing the turbulent viscosity functor");
+  params.addParam<bool>(
+      "add_dynamic_viscosity", true, "Whether add the dynamic viscosity in the expression");
```
Comment (GiudGiud) [inline modules/navier_stokes/src/functormaterials/FunctorEffectiveDynamicViscosity.C:33]:
```suggestion
      "add_dynamic_viscosity", true, "Whether to add the dynamic viscosity in the expression");
```

**Comment by GiudGiud** [comment]
Test failures are unrelated

File: framework/src/physics/PhysicsComponentInterface.C [hunk 18]
```diff
@@ -21,14 +21,16 @@ PhysicsComponentInterface::validParams()
...
+  // addRequiredPhysicsTask("add_ic");
 }
 
 void
 PhysicsComponentInterface::actOnAdditionalTasks()
 {
   // TODO: find a way to force this routine to be called by derived classes
-  if (_current_task == "add_ic")
+  // TODO: only keep the ics_physics task to prevent the risk of calling this twice
+  if ((_current_task == "add_ic" || _current_task == "add_ics_physics"))
```
Comment (grmnptr) [inline framework/src/physics/PhysicsComponentInterface.C:33]: Is this fix piggybacking on this PR?

File: modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C [hunk 19]
```diff
@@ -89,7 +89,8 @@ LinearFVTurbulentDiffusion::computeBoundaryMatrixContribution(const LinearFVBoun
   if (!diff_bc->includesMaterialPropertyMultiplier())
   {
     const auto face_arg = singleSidedFaceArg(_current_face_info);
-    grad_contrib *= _diffusion_coeff(face_arg, determineState());
+    grad_contrib *=
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C:92]: If we have the correct functor for computing the effective viscosity, can we just ge trid of this factor in this whole kernel in general? I wonder if that coul enable just overriding the addition kernels and keeping the general diffusion compute kernels.

File: modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C [hunk 20]
```diff
@@ -139,13 +139,15 @@ WCNSLinearFVFluidHeatTransferPhysics::addEnergyHeatConductionKernels()
...
+          _thermal_conductivity_name[block_i] + (_has_turbulence_model ? "_plus_kt" : "");
     }
     else
     {
       params.set<LinearVariableName>("variable") = _fluid_enthalpy_name;
-      params.set<MooseFunctorName>("diffusion_coeff") =
-          _thermal_conductivity_name[block_i] + "_by_cp";
+      const auto th_cond_name =
+          _thermal_conductivity_name[block_i] + (_has_turbulence_model ? "_plus_kt" : "");
+      params.set<MooseFunctorName>("diffusion_coeff") = th_cond_name + "_by_cp";
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:150]: We could also call this specific thermal diffusivity.

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+      true,
+      "Whether to use a non-orthogonal correction. This can potentially slow down convergence "
+      ", but reduces numerical dispersion on non-orthogonal meshes. Can be safely turned off on "
+      "orthogonal meshes.");
+  params.addParamNamesToGroup("use_nonorthogonal_correction", "Numerical scheme");
+  return params;
+}
+
+WCNSLinearFVTurbulencePhysics::WCNSLinearFVTurbulencePhysics(const InputParameters & parameters)
+  : WCNSFVTurbulencePhysicsBase(parameters)
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:65]: At some point it would be nice to create different physics for the different flavors of turbulence. MixingLenghtTurbulencePhysics, KEpsilonturbulence physics etc. Maybe not for now just sometime in the future.

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+    const std::string mat_type = "FunctorEffectiveDynamicViscosity";
+    InputParameters params = getFactory().getValidParams(mat_type);
+    assignBlocks(params, _blocks);
+    params.set<MooseFunctorName>("property_name") = "mu_eff_tked";
+    params.set<MooseFunctorName>(NS::mu) = _flow_equations_physics->dynamicViscosityName();
+    params.set<MooseFunctorName>(NS::mu_t) = _turbulent_viscosity_name;
+    params.set<MooseFunctorName>(NS::mu_t + "_inverse_factor") =
+        getParam<MooseFunctorName>("sigma_eps");
+    getProblem().addMaterial(mat_type, prefix() + "mu_eff_tked", params);
+  }
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:310]:
```suggestion
  // Functor materials for the diffusion coefficients
  if (_turbulence_model == "k-epsilon")
  {
    // Since sigma_k = 1 in the standard k-epsilon, this is often unnecessary
    const std::string mat_type = "FunctorEffectiveDynamicViscosity";
    InputParameters params = getFactory().getValidParams(mat_type);
    assignBlocks(params, _blocks);
    params.set<MooseFunctorName>("property_name") = "mu_eff_tke";
    params.set<MooseFunctorName>(NS::mu) = _flow_equations_physics->dynamicViscosityName();
    params.set<MooseFunctorName>(NS::mu_t) = _turbulent_viscosity_name;
    params.set<MooseFunctorName>(NS::mu_t + "_inverse_factor") =
        getParam<MooseFunctorName>("sigma_k");
    getProblem().addMaterial(mat_type, prefix() + "mu_eff_tke", params);
  
    params.set<MooseFunctorName>("property_name") = "mu_eff_tked";
    params.set<MooseFunctorName>(NS::mu_t + "_inverse_factor") =
        getParam<MooseFunctorName>("sigma_eps");
    getProblem().addMaterial(mat_type, prefix() + "mu_eff_tked", params);
  }
```

**Review by grmnptr — CHANGES_REQUESTED** [review]

File: framework/src/physics/PhysicsComponentInterface.C (hunk 18 above)
Comment (GiudGiud) [reply framework/src/physics/PhysicsComponentInterface.C:33]: cant do the PR without it

File: modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C (hunk 19 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C:92]: yes by changing every single k-epsilon input we could

File: modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C (hunk 20 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:150]: ? isnt that thermal conductivity?

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 21 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:65]: I was thinking we might get away with a single class. It works fine for WCNSFV. I think having a second model might be the test of that

File: modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C (hunk 20 above)
Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:150]: https://en.wikipedia.org/wiki/Thermal_diffusivity

Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:150]: This is just something to consider, I am fine with k/cp as well.

File: modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C (hunk 19 above)
Comment (grmnptr) [reply modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C:92]: I think we should do it. I can help if needed. Less code duplication. And I think having one kernel should be the supported direction considering other directions might encounter failures.

Comment (GiudGiud) [reply modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C:92]: You are not going to be able to delete the kernel because of the boundary/wall treatment.

Comment (grmnptr) [reply modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C:92]: I am not proposing to delee the kernel. Just the `compute*` functions. 
The functions like `addMatrixContribution` can stay.

Comment (grmnptr) [reply modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C]: Okay I'll do it then.

File: modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C (hunk 14 above)
Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C]: Nice catch!

**Review by grmnptr — APPROVED** [review]
Looks good from my side. 
Will get someone for a technical review.

**Comment by grmnptr** [comment]
Added this:
https://github.com/idaholab/moose/issues/31952

**Comment by tanoret** [comment]
Technical Review



**CASE 1: Channel ERCOFTAC 32 - Re_tau = 395**

Convergence:
Non-physics: 3730 iterations
Physics: 3724 --> OK

Non-physics: TOP; Physics: BOTTOM

Pressure --> OK
image


y_plus --> OK
image

TKE --> OK
image


TKED --> OK
image


**Test: PASS**


**CASE 2: Channel ERCOFTAC 32 - Re_tau = 590**

Convergence:
Non-physics: 2926 iterations
Physics: 2917 --> OK

Non-physics: TOP; Physics: BOTTOM

Pressure --> OK
image

y_plus --> OK
image

TKE --> OK
image

TKED --> OK
image

**Test: PASS**


**CASE 3: Channel ERCOFTAC 30 BFS**

In progress...

**Comment by tanoret** [comment]
Technical Review Continued

**CASE 3: BFS**

Convergence with physics:
Iteration 6000 Initial residual norms:
 Momentum equation: Component 1 2.41151e-07 Linear its: 1
 Momentum equation: Component 2 1.01668e-06 Linear its: 1
 Pressure equation: 2.93526e-07 Linear its: 2
 Advected system: TKE_system 1.43695e-06 Linear its: 1
 Advected system: TKED_system 1.93788e-06 Linear its: 1

Convergence without physics:
Iteration 6000 Initial residual norms:
 Momentum equation: Component 1 2.44858e-07 Linear its: 1
 Momentum equation: Component 2 1.08174e-06 Linear its: 2
 Pressure equation: 1.85106e-07 Linear its: 2
 Advected system: TKE_system 1.77218e-06 Linear its: 1
 Advected system: TKED_system 2.26245e-06 Linear its: 1 --> OK

Post-processing: Top -> non-physics; Bottom -> Physics

Pressure --> OK
image

y_plus --> OK
image

TKE --> OK
image

TKED --> OK
image

**Status: PASS**

**Comment by tanoret** [comment]
@GiudGiud The technical review is done. All good!

The cases with physics are in this PR: https://github.com/idaholab/open_pronghorn/pull/46

**Comment by GiudGiud** [comment]
Thanks for the review!
A libmesh update will go in first, then this can go

**Comment by grmnptr** [comment]
Technical review is here:
https://github.com/idaholab/open_pronghorn/pull/46

**Comment by grmnptr** [comment]
Aaa sorry, I didn't see the details above. It is good to merge from my side, the PR is ready on OpenPronghorn.

**Comment by GiudGiud** [comment]
thanks for the reviews guys!
Let's get your PRs in next Peter
